frontend/src/query_builder/ExpressionWidget.jsx

```diff
@@ -0,0 +1,465 @@
+import React, { Component, PropTypes } from 'react';
+
+import _ from 'underscore';
+
+import Icon from "metabase/components/Icon.jsx";
+import Popover from "metabase/components/Popover.jsx";
+
+
+const VALID_OPERATORS = new Set(['+', '-', '*', '/']);
+
+const OPERATOR_SUGGESTIONS = [
+    {name: '+'},
+    {name: '-'},
+    {name: '*'},
+    {name: '/'}
+];
+
+const ERROR_MESSAGE_EMPTY_NAME       = 'what should this expression be called?';
+const ERROR_MESSAGE_EMPTY_EXPRESSION = 'enter an expression';
+
+const KEYCODE_ENTER = 13;
+const KEYCODE_UP    = 38;
+const KEYCODE_DOWN  = 40;
+
+
+function tokenizeExpression(expression, i = 0, level = 0) {
```

Comment:
all of the functions and constants which are not actually part of the react component display are best moved into a separate file.  you can put them in query.js inside something like `ExpressionClause` if that makes sense, or possibly in their own file, up to you.  and if possible lets add some unit tests for functions like this which are easily able to be tested.